File: About-Swift-Advanced/NLFoo.swift
import Foundation

class NLFoo {
  private let ref: NLFooRef
  
  init(value: Int32) {
    ref = NLFooCreate(value)
  }
  
  var value: Int32 {
    get {
      return NLFooGetValue(ref)
    }
    set {
      NLFooSetValue(ref, newValue)
    }
  }
  
  deinit {
    NLFooRelease(ref)
  }
}

File: About-Swift-Advanced/NLFoo.c
#include "NLFoo.h"

struct NLFoo {
  int value;
};

NLFooRef NLFooCreate(int value)
{
  NLFooRef foo = calloc(1, sizeof(struct NLFoo));
  foo->value = value;
  return foo;
}

int NLFooGetValue(NLFooRef foo)
{
  return foo->value;
}

void NLFooSetValue(NLFooRef foo, int value)
{
  foo->value = value;
}

void NLFooRelease(NLFooRef foo)
{
  free(foo);
}
